### PR TITLE
[extractor/triller] Add Triller extractor

### LINKED ISSUE #4703: [Feature Request] Add support for Triller
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.08.19** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

United States

### Example URLs

- Account: https://triller.co/@charlidamelio
- Single video: https://triller.co/@charlidamelio/video/46c6fcfa-aa9e-4503-a50c-68444f44cddc
- 
On each page load, a file called `media.ts` is downloaded, this seems to contain the full video. Example URL: https://uploads.cdn.triller.co/v1/uploads/71621277/hls/media-5/media.ts

### Provide a description that is worded well enough to be understood

(I'm not sure what goes here)

[Triller](https://en.wikipedia.org/wiki/Triller_(app)) is a short video website, very similar to TikTok. `yt-dlp` already supports Tiktok, so I figured it would make sense to also support Triller.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-v', 'https://triller.co/@charlidamelio']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.08.19 [48c88e0] (zip)
[debug] Python 3.10.4 (CPython 64bit) - Linux-5.15.0-46-generic-x86_64-with-glibc2.35 (glibc 2.35)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2
[debug] Optional libraries: brotli-1.0.9, certifi-2020.06.20, secretstorage-3.3.1, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] [generic] Extracting URL: https://triller.co/@charlidamelio
[generic] @charlidamelio: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] @charlidamelio: Extracting information
[debug] Looking for Brightcove embeds
[debug] Looking for embeds
ERROR: Unsupported URL: https://triller.co/@charlidamelio
Traceback (most recent call last):
  File "/home/raleigh/Downloads/./yt-dlp/yt_dlp/YoutubeDL.py", line 1442, in wrapper
    return func(self, *args, **kwargs)
  File "/home/raleigh/Downloads/./yt-dlp/yt_dlp/YoutubeDL.py", line 1518, in __extract_info
    ie_result = ie.extract(url)
  File "/home/raleigh/Downloads/./yt-dlp/yt_dlp/extractor/common.py", line 666, in extract
    ie_result = self._real_extract(url)
  File "/home/raleigh/Downloads/./yt-dlp/yt_dlp/extractor/generic.py", line 3077, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://triller.co/@charlidamelio
```

### PR DESCRIPTION
**Pull request by bashonly**
</details>

Adds support for triller.co, a short-form video social media platform

Closes #4703 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by raleighlittles** [comment]
Hey, I'm the person that created the Triller feature request (#4703). I just wanted to say thank you for implementing this, and so quickly at that. I really appreciate it.

File: yt_dlp/extractor/triller.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+        username = user_info.get('username')
+        formats = []
+        video_url = video_info.get('video_url') or video_info.get('stream_url')
+        if video_url:
+            formats.append({
+                'url': video_url,
+                'ext': 'mp4',
+                'vcodec': 'h264',
+                'resolution': f'{video_info.get("width")}x{video_info.get("height")}',
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:106]:
```suggestion
```
core code will set this

File: yt_dlp/extractor/triller.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+            formats.append({
+                'url': video_url,
+                'ext': 'mp4',
+                'vcodec': video.get('codec'),
+                'vbr': bitrate,
+                'resolution': resolution,
+                'width': int_or_none(resolution.split('x')[0]),
+                'height': int_or_none(resolution.split('x')[1]),
+                'format_id': url_basename(video_url).split('.')[0],
+            })
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:126]:
```suggestion
            resolution = video.get('resolution') or ''
            formats.append({
                'url': video['url'],
                'ext': 'mp4',
                'vcodec': video.get('codec'),
                'vbr': int_or_none(video.get('bitrate'), 1000),
                'width': int_or_none(resolution.split('x')[0]),
                'height': int_or_none(resolution.split('x')[1]),
                'format_id': url_basename(video['url']).split('.')[0],
            })
```

File: yt_dlp/extractor/triller.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+            'duration': int_or_none(video_info.get('duration')),
+            'view_count': int_or_none(video_info.get('play_count')),
+            'like_count': int_or_none(video_info.get('likes_count')),
+            'artist': str_or_none(video_info.get('song_artist')),
+            'track': str_or_none(video_info.get('song_title')),
+            'webpage_url': f'https://triller.co/@{username}/video/{video_uuid}',
+            'uploader_url': f'https://triller.co/@{username}',
+            'formats': formats,
+            'comment_count': comment_count,
+            'comments': comments,
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:164]:
```suggestion
            '__post_extractor': self.extract_comments(video_id, comment_count),
```

File: yt_dlp/extractor/triller.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+            'creator': str_or_none(user_info.get('name')),
+            'timestamp': unified_timestamp(video_info.get('timestamp')),
+            'upload_date': unified_strdate(video_info.get('timestamp')),
+            'duration': int_or_none(video_info.get('duration')),
+            'view_count': int_or_none(video_info.get('play_count')),
+            'like_count': int_or_none(video_info.get('likes_count')),
+            'artist': str_or_none(video_info.get('song_artist')),
+            'track': str_or_none(video_info.get('song_title')),
+            'webpage_url': f'https://triller.co/@{username}/video/{video_uuid}',
+            'uploader_url': f'https://triller.co/@{username}',
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:161]:
```suggestion
            'uploader_url': format_field(username, None, 'https://triller.co/@%s'),
```

File: yt_dlp/extractor/triller.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+        } for comment_dict in comment_data] if comment_data else None
+        return comments
+
+    def _parse_video_info(self, video_info, user_info=None):
+        video_uuid = video_info.get('video_uuid')
+        video_id = video_info.get('id')
+        if not user_info:
+            user_info = traverse_obj(video_info, 'user', default={})
+
+        username = user_info.get('username')
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:98]: Move near where it is used

File: yt_dlp/extractor/triller.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+            'artist': 'Dixie',
+            'track': 'Someone to Blame',
+            'webpage_url': 'https://triller.co/@charlidamelio/video/46c6fcfa-aa9e-4503-a50c-68444f44cddc',
+            'uploader_url': 'https://triller.co/@charlidamelio',
+            'comment_count': int,
+        }
+    }]
+
+    def _real_extract(self, url):
+        username, video_uuid = self._match_valid_url(url).groups()
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:227]:
```suggestion
        username, video_uuid = self._match_valid_url(url).group('username', 'id')
```

File: yt_dlp/extractor/triller.py [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+
+        video_info = traverse_obj(self._download_json(
+            f'{self._API_BASE_URL}/api/videos/{video_uuid}',
+            video_uuid, note='Downloading video info API JSON',
+            errnote='Unable to download video info API JSON',
+            headers={
+                'Origin': 'https://triller.co',
+            }), ('videos', 0))
+        if not video_info:
+            raise ExtractorError('No video info found in API response')
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:237]: is this `expected=True`?

File: yt_dlp/extractor/triller.py [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+        if user_info.get('private') and not user_info.get('followed_by_me'):
+            raise ExtractorError('This video is private', expected=True)
+
+        if user_info.get('blocked_by_user') or user_info.get('blocking_user'):
+            raise ExtractorError('The author of the video is blocked', expected=True)
+
+        return {
+            **self._parse_video_info(video_info, user_info),
+        }
+
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:252]:
```suggestion
        user_info = video_info.get('user') or {}
        if not user_info:
            self.report_warning('Unable to extract user info')
        elif user_info.get('private') and not user_info.get('followed_by_me'):
            raise ExtractorError('This video is private', expected=True)
        elif traverse_obj(user_info, 'blocked_by_user', 'blocking_user'):
            raise ExtractorError('The author of the video is blocked', expected=True)

        return self._parse_video_info(video_info, user_info)

```

File: yt_dlp/extractor/triller.py [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+                            'Origin': 'https://triller.co',
+                        }, query=query)
+                except ExtractorError as e:
+                    if isinstance(e.cause, json.JSONDecodeError) and e.cause.pos == 0:
+                        retry.error = e
+                        continue
+                    raise
+            yield from video_list.get('videos', [])
+            if not video_list.get('videos'):
+                break
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:296]:
```suggestion
            if not video_list.get('videos'):
                break
            yield from video_list['videos']
```

File: yt_dlp/extractor/triller.py [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+            before_time = traverse_obj(video_list, ('videos', -1, 'timestamp'))
+            if not before_time:
+                break
+            query = {
+                'before_time': before_time,
+                'limit': limit,
+            }
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:303]:
```suggestion
            query['before_time'] = traverse_obj(video_list, ('videos', -1, 'timestamp'))
            if not query['before_time']:
                break
```

File: yt_dlp/extractor/triller.py [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+        for video in videos:
+            yield {
+                **self._parse_video_info(video, user_info),
+                'extractor_key': TrillerIE.ie_key(),
+                'extractor': 'Triller',
+            }
+
+    def _real_extract(self, url):
+        username = self._match_id(url)
+        if not self._IS_LOGGED_IN:
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:315]:
```suggestion
        if not self._AUTH_TOKEN:
```
_IS_LOGGED_IN is not even needed as far as I can tell

File: yt_dlp/extractor/triller.py [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+            None, note='Creating guest session', data=b'', headers={
+                'Origin': 'https://triller.co',
+            }, query={
+                'platform': 'Web',
+                'app_version': '',
+            })
+        if not guest.get('auth_token'):
+            raise ExtractorError('Unable to fetch required auth token for user extraction')
+
+        self._AUTH_TOKEN = guest.get('auth_token')
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:71]:
```suggestion
        self._AUTH_TOKEN = guest['auth_token']
```

File: yt_dlp/extractor/triller.py (hunk 11 above)
Comment (pukkandan) [reply yt_dlp/extractor/triller.py:315]: Also, this should be in _real_initialize imo

File: yt_dlp/extractor/triller.py [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+                'Origin': 'https://triller.co',
+            }).get('user')
+        if not user_info:
+            raise ExtractorError('Unable to extract user info')
+
+        if user_info.get('private') and not user_info.get('followed_by_me'):
+            raise ExtractorError('This user profile is private', expected=True)
+
+        if user_info.get('blocked_by_user') or user_info.get('blocking_user'):
+            raise ExtractorError('This user profile is blocked', expected=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:332]: repeated code from above. Refactor into a function

File: yt_dlp/extractor/triller.py [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+                'limit': limit,
+            }
+
+    def _entries(self, videos, user_info):
+        for video in videos:
+            yield {
+                **self._parse_video_info(video, user_info),
+                'extractor_key': TrillerIE.ie_key(),
+                'extractor': 'Triller',
+            }
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:311]: This can just be in `_parse_video_info`, no?

File: yt_dlp/extractor/triller.py [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+                'before_time': before_time,
+                'limit': limit,
+            }
+
+    def _entries(self, videos, user_info):
+        for video in videos:
+            yield {
+                **self._parse_video_info(video, user_info),
+                'extractor_key': TrillerIE.ie_key(),
+                'extractor': 'Triller',
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:310]:
```suggestion
                'extractor': TrillerIE.IE_NAME,
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/triller.py (hunk 11 above)
Comment (bashonly) [reply yt_dlp/extractor/triller.py]: yeah `_IS_LOGGED_IN` is redundant, removed

moved `_create_guest` into the `_real_initialize` function for `TrillerUserIE` -- an auth token is only required when scraping user profiles, not individual videos

File: yt_dlp/extractor/triller.py (hunk 14 above)
Comment (bashonly) [reply yt_dlp/extractor/triller.py]: yes

File: yt_dlp/extractor/triller.py (hunk 7 above)
Comment (bashonly) [reply yt_dlp/extractor/triller.py:222]: no. should it be removed then?

File: yt_dlp/extractor/triller.py [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+        self._AUTH_TOKEN = login['auth_token']
+
+    def _get_comments(self, video_id, limit=15):
+        comment_info = self._download_json(
+            f'{self._API_BASE_URL}/api/videos/{video_id}/comments_v2',
+            video_id, fatal=False, note=False, errnote='Unable to extract comments',
+            headers={'Origin': 'https://triller.co'}, query={'limit': limit}) or {}
+        if not comment_info.get('comments'):
+            return
+        yield [{
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:60]:
```suggestion
        return [{
```

File: yt_dlp/extractor/triller.py [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+            formats.append({
+                'url': video_url,
+                'ext': 'mp4',
+                'vcodec': 'h264',
+                'width': video_info.get('width'),
+                'height': video_info.get('height'),
+                'format_id': url_basename(video_url).split('.')[0],
+                'filesize': video_info.get('filesize'),
+            })
+        video_set = video_info.get('video_set', [])
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:93]:
```suggestion
        video_set = video_info.get('video_set') or []
```

File: yt_dlp/extractor/triller.py [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+            formats.extend(self._extract_m3u8_formats(
+                manifest_url, video_id, 'mp4', entry_protocol='m3u8_native',
+                m3u8_id='hls', fatal=False, note=False))
+        self._sort_formats(formats)
+
+        comment_count = int_or_none(video_info.get('comment_count'))
+
+        if not user_info:
+            user_info = traverse_obj(video_info, 'user', default={})
+
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:124]:
```suggestion
        user_info = user_info or traverse_obj(video_info, 'user', default={})

```

File: yt_dlp/extractor/triller.py [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+            'uploader_id': str_or_none(user_info.get('user_id')),
+            'creator': str_or_none(user_info.get('name')),
+            'timestamp': unified_timestamp(video_info.get('timestamp')),
+            'upload_date': unified_strdate(video_info.get('timestamp')),
+            'duration': int_or_none(video_info.get('duration')),
+            'view_count': int_or_none(video_info.get('play_count')),
+            'like_count': int_or_none(video_info.get('likes_count')),
+            'artist': str_or_none(video_info.get('song_artist')),
+            'track': str_or_none(video_info.get('song_title')),
+            'webpage_url': f'https://triller.co/@{username}/video/{video_uuid}',
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:142]: This makes `username` a mandatory field. Is that intentional?

File: yt_dlp/extractor/triller.py [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+            video_uuid, note='Downloading video info API JSON',
+            errnote='Unable to download video info API JSON',
+            headers={
+                'Origin': 'https://triller.co',
+            }), ('videos', 0))
+        if not video_info:
+            raise ExtractorError('No video info found in API response')
+
+        user_info = self._check_user_info(video_info.get('user', {}))
+
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:224]:
```suggestion
        user_info = self._check_user_info(video_info.get('user') or {})
```

File: yt_dlp/extractor/triller.py [hunk 21]
```diff
@@ -0,0 +1,309 @@
...
+        user_info = self._check_user_info(self._download_json(
+            f'{self._API_BASE_URL}/api/users/by_username/{username}',
+            username, note='Downloading user info',
+            errnote='Failed to download user info', headers={
+                'Authorization': f'Bearer {self._AUTH_TOKEN}',
+                'Origin': 'https://triller.co',
+            }).get('user', {}))
+
+        user_id = str_or_none(user_info.get('user_id'))
+        videos = LazyList(self._extract_video_list(username, user_id))
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:305]:
```suggestion
        videos = self._extract_video_list(username, user_id)
```
why LazyList?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/triller.py (hunk 16 above)
Comment (bashonly) [reply yt_dlp/extractor/triller.py:60]: --write-comments throws an error with return instead of yield ?

File: yt_dlp/extractor/triller.py (hunk 19 above)
Comment (bashonly) [reply yt_dlp/extractor/triller.py:141]: `username` is always capturable via the url

Comment (pukkandan) [reply yt_dlp/extractor/triller.py:141]: but you aren't doing that. See L124

Comment (bashonly) [reply yt_dlp/extractor/triller.py:138]: good catch, thanks. fixed

File: yt_dlp/extractor/triller.py [hunk 22]
```diff
@@ -0,0 +1,303 @@
...
+            }, data=json.dumps({'username': username}, separators=(',', ':')).encode('utf-8'))
+        if user_check.get('status'):  # endpoint returns "status":false if username exists
+            raise ExtractorError('Unable to login: Invalid username', expected=True)
+
+        credentials = {
+            'username': username,
+            'password': password,
+        }
+        login = self._download_json(
+            f'{self._API_BASE_URL}/user/auth', None, note=False, fatal=False,
```
Comment (pukkandan) [inline yt_dlp/extractor/triller.py:39]: Don't abuse noe=False

**Pull request merged**